copilot_proxy/utils/codegen.py

```diff
@@ -72,14 +74,17 @@ def to_word_list_format(word_dict, tokenizer):
     def generate(self, data):
         model_name = "fastertransformer"
         prompt = data['prompt']
+        max_tokens = data.get('max_tokens', 16)
+        encoded_tokens = self.tokenizer.encode(prompt).ids
+        if max_tokens + len(encoded_tokens) > self.MAX_MODEL_LEN:
+            print("Max tokens + prompt length exceeds maximum model length, shortening prompt")
+            encoded_tokens = prompt_safety_valve(tokens=encoded_tokens, max_tokens=max_tokens,
```

Comment:
Overall I'm not sure about the approach of truncating prompts server-side rather than returning an error. The OpenAI API returns a 400 status that looks like:

```
InvalidRequestError: This model's maximum context length is 2049 tokens, however you requested 23582 tokens (23566 in your prompt; 16 for the completion). Please reduce your prompt; or completion length.                                                       
```

So we might want to take that approach for compatibility with their API, and instead fix overly-long prompts in the client code.

Reply:
I like your approach better, stick to the OpenAI spec as it will improve functionality with existing plugins 😄 I'll update it and do a force push.